File: leetcode/easy/twoSumIV.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    bool dfs(TreeNode* x, int k, unordered_set<int>& s) {
        if (x == nullptr) return false;
        if (s.count(k-x->val)) return true;
        s.insert(x->val);
        return (dfs(x->left, k, s) || dfs(x->right, k, s))
    }

    bool findTarget(TreeNode* root, int k) {
        unordered_set<int> s;
        return dfs(root, k, s);
    }
};

File: leetcode/easy/two_sum_IV.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: TreeNode, k: int) -> bool:
        s = set()
        def dfs(x):
            if not x: return False
            if k-x.val in s: return True
            s.add(x.val)
            return dfs(x.left) or dfs(x.right)
        return dfs(root)
